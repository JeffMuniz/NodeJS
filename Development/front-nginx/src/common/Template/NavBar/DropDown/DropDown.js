import React from "react";
import { oneOfType, func, shape, bool, number, string } from "prop-types";

import { WebPaths, Routes } from "src/routes/consts";
import { ClickOutsideListener } from "@common";
import { If } from "@utils";
import { dropDownMenu, deliveryTypes } from "@enums";

import {
  NavLinkStyled,
  DropDownWrapper,
  activeLink,
  BaseOption as LogoutOption,
  ChangeGroupOption,
} from "./DropDown.styles";

export const DropDown = ({
  triggerRef,
  isOpen,
  handleClickOutside,
  handleLogout,
  groupListLength,
  showChangeGroup,
  handleGroupModal,
  deliveryType,
}) => (
  <ClickOutsideListener
    id="ud_dropdown"
    handleClickOutside={handleClickOutside}
    isListening={isOpen}
    triggerRef={triggerRef}
  >
    <DropDownWrapper isOpen={isOpen}>
      {dropDownMenu
        .filter(link => {
          if (deliveryType === deliveryTypes.door2Door) {
            return link.route !== Routes.DELIVERY_PLACES;
          }

          return link;
        })
        .map(link => (
          <NavLinkStyled
            id={`ud_link_${link.route}`}
            key={link.route}
            exact
            onClick={handleClickOutside}
            to={WebPaths[link.route]}
            activeStyle={activeLink}
          >
            {link.label}
          </NavLinkStyled>
        ))}
      <If test={showChangeGroup}>
        <ChangeGroupOption onClick={handleGroupModal} id="btn_change_group">
          Trocar Grupo ({groupListLength})
        </ChangeGroupOption>
      </If>
      <LogoutOption id="logout_dropdown_button" onClick={handleLogout}>
        Sair
      </LogoutOption>
    </DropDownWrapper>
  </ClickOutsideListener>
);

export default DropDown;

DropDown.propTypes = {
  triggerRef: oneOfType([func, shape({})]).isRequired,
  isOpen: bool,
  handleClickOutside: func,
  handleLogout: func.isRequired,
  groupListLength: number,
  showChangeGroup: bool,
  handleGroupModal: func,
  deliveryType: string.isRequired,
};

DropDown.defaultProps = {
  isOpen: false,
  handleClickOutside: () => null,
  groupListLength: 0,
  showChangeGroup: false,
  handleGroupModal: () => null,
};
