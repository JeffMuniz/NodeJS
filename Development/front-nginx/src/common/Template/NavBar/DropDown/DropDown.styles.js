import styled from "styled-components";
import { NavLink } from "react-router-dom";
import { lighterBlack, white, darkGrey } from "src/styles/colors";
import { ifStyle } from "@utils";

const isOpen = ifStyle("isOpen");

export const NavLinkStyled = styled(NavLink)`
  color: ${lighterBlack};
  border-bottom: 1px solid ${darkGrey};
  font-size: 16px;
  padding: 15px 0;
  text-decoration: none;
  line-height: 26px;
`;

export const activeLink = {
  fontWeight: "bold",
};

export const BaseOption = styled.li`
  cursor: pointer;
  list-style-type: none;
  padding: 17.5px 0;
`;

export const DropDownWrapper = styled.div`
  background-color: ${white};
  box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);
  display: flex;
  flex-direction: column;
  opacity: ${isOpen("1", "0")};
  padding: 5px 24px 10px 24px;
  position: absolute;
  right: 0;
  transition: opacity 0.25s ease-in-out;
  z-index: 4;
  width: 175px;
  visibility: ${isOpen("visible", "hidden")};
`;

export const ChangeGroupOption = styled(BaseOption)`
  border-bottom: 1px solid ${darkGrey};
`;
