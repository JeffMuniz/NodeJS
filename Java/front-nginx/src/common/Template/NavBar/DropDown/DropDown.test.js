import React from "react";
import ShallowRenderer from "react-test-renderer/shallow";
import DropDown from "./DropDown";

jest.mock("@common", () => ({
  SvgIcon: "SvgIcon",
  ClickOutsideListener: "ClickOutsideListener",
}));

describe("DropDown - Component", () => {
  it("should render it dropdown", () => {
    // given
    const renderer = new ShallowRenderer();
    const props = {
      handleClickOutside: jest.fn(),
      triggerRef: { current: jest.fn() },
      handleLogout: jest.fn(),
      deliveryType: "HR",
    };

    // when
    renderer.render(<DropDown {...props} />);
    const result = renderer.getRenderOutput();

    // then
    expect(result).toMatchSnapshot();
  });
});
