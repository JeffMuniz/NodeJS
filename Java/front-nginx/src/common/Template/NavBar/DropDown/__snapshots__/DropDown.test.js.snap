// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDown - Component should render it dropdown 1`] = `
<ClickOutsideListener
  handleClickOutside={[MockFunction]}
  id="ud_dropdown"
  isListening={false}
  triggerRef={
    Object {
      "current": [MockFunction],
    }
  }
>
  <styled.div
    isOpen={false}
  >
    <Styled(NavLink)
      activeStyle={
        Object {
          "fontWeight": "bold",
        }
      }
      exact={true}
      id="ud_link_RhFront.AccessPermission"
      onClick={[MockFunction]}
      to="/permissao-accesso"
    >
      Permiss√£o de Acesso
    </Styled(NavLink)>
    <Styled(NavLink)
      activeStyle={
        Object {
          "fontWeight": "bold",
        }
      }
      exact={true}
      id="ud_link_RhFront.FileConverter"
      onClick={[MockFunction]}
      to="/conversor"
    >
      Converter Arquivos
    </Styled(NavLink)>
    <Styled(NavLink)
      activeStyle={
        Object {
          "fontWeight": "bold",
        }
      }
      exact={true}
      id="ud_link_RhFront.DeliveryPlaces"
      onClick={[MockFunction]}
      to="/locais-entrega"
    >
      Locais de Entrega
    </Styled(NavLink)>
    <If
      test={false}
    >
      <Styled(styled.li)
        id="btn_change_group"
        onClick={[Function]}
      >
        Trocar Grupo (
        0
        )
      </Styled(styled.li)>
    </If>
    <styled.li
      id="logout_dropdown_button"
      onClick={[MockFunction]}
    >
      Sair
    </styled.li>
  </styled.div>
</ClickOutsideListener>
`;
